### PR TITLE
[FSDP] Remove unneeded padding logic for optim state dict

### PR DESCRIPTION
**Pull request by awgu**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #78209 [Do Not Merge] Test if 4 GPU test runs
* #78224 [FSDP] Add optim state dict shape check
* **#78208 [FSDP] Remove unneeded padding logic for optim state dict**

Differential Revision: [D36643992](https://our.internmc.facebook.com/intern/diff/D36643992)

### CODE REVIEW

File: torch/distributed/fsdp/_optim_utils.py [hunk 1]
```diff
@@ -497,26 +499,11 @@ def _flatten_tensor_optim_state(
         for state_value, shape
         in zip(pos_dim_tensors, unflat_param_shapes)
     ]
-    padding = flat_param.num_padded
```
Comment (awgu) [inline torch/distributed/fsdp/_optim_utils.py:500]: Adding this padding was unnecessary. `flat_param.num_padded` is the number of elements padded for this local rank's shard of the flattened parameter.

`flat_tensor` is the concatenation of all flattened unsharded optimizer state tensors, which means it has the same shape as the flattened unsharded parameter without padding. (See the new assert on L504.)

If we pad `flat_tensor` by any amount between 0 and the total amount of padding for the flattened parameter, then the padded `flat_tensor` will still shard correctly. All it means is that the amount of padding added in `FullyShardedDataParallel._get_shard()` will be less (i.e. we pad some here and pad some in `_get_shard()` instead of padding all in `_get_shard()`).

Because `flat_param.num_padded` satisfies that inequality of being between 0 and the total amount of padding, this did not cause any correctness issues. However, because this is confusing since we pad the unsharded `flat_tensor` by possibly a different amount per rank, I am removing this logic.

File: test/distributed/fsdp/test_fsdp_optim_state.py [hunk 2]
```diff
@@ -446,8 +446,7 @@ def test_full_optim_state_dict_nested(
             full_osd, ref_osd, check_same_param_keys=check_same_param_keys,
         )
 
-    # Require 4 GPUs since we test halving the world size
-    @skip_if_lt_x_gpu(4)
+    @skip_if_lt_x_gpu(2)
```
Comment (awgu) [inline test/distributed/fsdp/test_fsdp_optim_state.py:449]: To make sure these tests always run, we decrease the GPU requirement to 2. This means that the tests that halve the world size may halve from 2 to 1, in which case the parameters are no longer sharded. This is still a meaningful exercise of the saving and loading.

The halving of the world size is done by only including even ranks, meaning that the test's implementation is adaptable to different total number of GPUs (e.g. whether 2 or 4).

**Comment by awgu** [comment]
@awgu has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D36643992).

File: torch/distributed/fsdp/_optim_utils.py (hunk 1 above)
Comment (rohan-varma) [reply torch/distributed/fsdp/_optim_utils.py:500]: Sounds good! Does FSDP._get_shard() mention that it handles padding in the documentation so users of the API don't have to (and probably shouldn't) handle padding themselves?

**Review by rohan-varma — APPROVED** [review]
LGTM, thanks!

File: torch/distributed/fsdp/_optim_utils.py (hunk 1 above)
Comment (awgu) [reply torch/distributed/fsdp/_optim_utils.py:500]: We can clarify that the input tensor to `_get_shard()` does not need to be padded and that the method will pad it.

**Comment by awgu** [comment]
Failures look unrelated.
